### PR TITLE
TorchFX Dynamic Shapes Additional Support

### PR DESCRIPTION
**Pull request by cavusmustafa**
### Details:
 - Reshape dynamic input support updated to use Concat if there are unkown dimensions
 - Dynamic shape support added for arange and full ops. 

### Tickets:
 - *ticket-id*
 -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -

### CODE REVIEW

File: src/frontends/pytorch/src/op/slice_scatter.cpp [hunk 1]
```diff
@@ -44,7 +44,7 @@ OutputVector translate_slice_scatter_fx(const NodeContext& context) {
     ov::Output<ov::Node> end;
     if (!context.input_is_none(4)) {
         end = context.get_input(4);
-        if (end.get_partial_shape().rank().is_dynamic() || end.get_partial_shape().rank().get_length() == 0) {
+        if (!(end.get_partial_shape().rank().is_dynamic()) && end.get_partial_shape().rank().get_length() == 0) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/slice_scatter.cpp:47]: Why this change? And why other inputs are unchanged?

File: src/frontends/pytorch/src/op/arange.cpp [hunk 2]
```diff
@@ -108,6 +109,9 @@ OutputVector translate_arange_fx(const NodeContext& context) {
     if (context.has_attribute("dtype")) {
         dtype = context.get_attribute<element::Type>("dtype");
     }
+    if (end.get_partial_shape().rank().is_dynamic()) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/arange.cpp:112]: Why only in dynamic case? What if it is static and 1D?

File: src/frontends/pytorch/src/op/full.cpp [hunk 3]
```diff
@@ -74,9 +76,28 @@ OutputVector translate_full(const NodeContext& context) {
...
     // aten.full.default([16, 16], 0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'),
     // pin_memory = False)
-    num_inputs_check(context, 2, 2);
-    auto sizes = context.get_input(0);
-    auto value = context.get_input(1);
+    auto num_inputs = context.get_input_size();
+    num_inputs_check(context, 2, num_inputs);
+    ov::Output<ov::Node> sizes;
+    if (context.get_input_type(0).is<type::List>()) {
+        std::deque<Output<Node>> list_elems;
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/full.cpp:83]:
```suggestion
        OutputVector list_elems;
```

File: src/frontends/pytorch/src/op/full.cpp [hunk 4]
```diff
@@ -74,9 +76,28 @@ OutputVector translate_full(const NodeContext& context) {
...
+                std::vector<int32_t> dim_vec;
+                dim_vec.push_back(const_val);
+                auto dim_const = ov::op::v0::Constant::create(element::i32, Shape{1}, dim_vec);
+                list_elems.push_back(dim_const);
+            } else {
+                auto converted_dim = context.mark_node(std::make_shared<ov::op::v0::Convert>(context.get_input(static_cast<int>(i)), element::i32));
+                list_elems.push_back(converted_dim);
+            }
+        }
+        sizes = std::make_shared<ov::op::v0::Concat>(OutputVector(list_elems.begin(), list_elems.end()), 0);
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/full.cpp:96]:
```suggestion
        sizes = std::make_shared<ov::op::v0::Concat>(list_elems, 0);
```

File: src/frontends/pytorch/src/op/full.cpp [hunk 5]
```diff
@@ -74,9 +76,28 @@ OutputVector translate_full(const NodeContext& context) {
...
+    } else {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/full.cpp:97]: Can this be a helper function and be reused for full and reshape too?

Comment (cavusmustafa) [reply src/frontends/pytorch/src/op/full.cpp]: I created a helper function into utils.cpp for the main loop to create the concat op.

File: src/frontends/pytorch/src/op/arange.cpp (hunk 2 above)
Comment (cavusmustafa) [reply src/frontends/pytorch/src/op/arange.cpp]: Makes sense to add this change but we haven't observed such an input on the models we are working on. I updated this line to include static 1D inputs for end.

File: src/frontends/pytorch/src/op/slice_scatter.cpp (hunk 1 above)
Comment (cavusmustafa) [reply src/frontends/pytorch/src/op/slice_scatter.cpp:47]: This was initially changed due to a failure in Lama3. I updated rest of the input checks accordingly to make it consistent.

File: src/frontends/pytorch/src/op/arange.cpp [hunk 6]
```diff
@@ -108,6 +109,10 @@ OutputVector translate_arange_fx(const NodeContext& context) {
     if (context.has_attribute("dtype")) {
         dtype = context.get_attribute<element::Type>("dtype");
     }
+    if (end.get_partial_shape().rank().is_dynamic() ||
+        (end.get_partial_shape().rank().is_static() && end.get_partial_shape().rank().get_length() == 1)) {
+        end = context.mark_node(std::make_shared<ov::op::v0::Squeeze>(end, zero));
+    }
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/arange.cpp:115]: I don't like this change. You only do this for `end`, while `start` and `step` are similar. Could you make a test? `torch.export` test can cover this case.

File: src/frontends/pytorch/src/op/expand.cpp [hunk 7]
```diff
@@ -46,23 +48,38 @@ OutputVector translate_expand_fx(const NodeContext& context) {
     num_inputs_check(context, 2, num_inputs);
     auto x = context.get_input(0);
     std::vector<int32_t> shape_vec;
-    auto sizes = context.get_input(1);
-    if (num_inputs != 2) {
+    if (context.get_input_type(1).is<type::List>()) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/expand.cpp:51]: Can't you use `concat_dims_to_shape` here too?

File: src/frontends/pytorch/src/utils.cpp [hunk 8]
```diff
@@ -555,6 +555,25 @@ Output<Node> masked_fill(ov::pass::NodeRegistry& rg,
     return rg.make<opset10::Select>(bool_mask, _value, data);
 }
 
+Output<Node> concat_dims_to_shape(const NodeContext& context, size_t begin, size_t end) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.cpp:558]: Naming is not great. It can be used not only for shapes, but for any list inputs, I propose name `concat_list_from_inputs`

File: src/frontends/pytorch/src/utils.cpp [hunk 9]
```diff
@@ -555,6 +555,25 @@ Output<Node> masked_fill(ov::pass::NodeRegistry& rg,
...
+    OutputVector list_elems;
+    for (size_t i = begin; i < end; i++) {
+        if (context.get_input_type(i).as<type::List>().element_type.is<type::PyScalar>()) {
+            auto const_val = context.const_input<int32_t>(i);
+            std::vector<int32_t> dim_vec;
+            dim_vec.push_back(const_val);
+            auto dim_const = ov::op::v0::Constant::create(element::i32, Shape{1}, dim_vec);
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.cpp:565]: PyScalar is int64 type

File: src/frontends/pytorch/src/utils.cpp [hunk 10]
```diff
@@ -555,6 +555,25 @@ Output<Node> masked_fill(ov::pass::NodeRegistry& rg,
...
+            list_elems.push_back(dim_const);
+        } else {
+            auto converted_dim = context.mark_node(
+                std::make_shared<ov::op::v0::Convert>(context.get_input(static_cast<int>(i)), element::i32));
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.cpp:569]: Do you really need to convert inputs?
Also, why don't you unsqueeze inputs? If it was an element of list it should be unsqueezed, shouldn't it?

File: src/frontends/pytorch/src/op/expand.cpp (hunk 7 above)
Comment (cavusmustafa) [reply src/frontends/pytorch/src/op/expand.cpp:51]: Updated accordingly

File: src/frontends/pytorch/src/utils.cpp (hunk 8 above)
Comment (cavusmustafa) [reply src/frontends/pytorch/src/utils.cpp]: Updated the naming

File: src/frontends/pytorch/src/utils.cpp (hunk 9 above)
Comment (cavusmustafa) [reply src/frontends/pytorch/src/utils.cpp]: changed to use int64

File: src/frontends/pytorch/src/utils.cpp (hunk 10 above)
Comment (cavusmustafa) [reply src/frontends/pytorch/src/utils.cpp]: Removed the convert inputs. Unsqueeze inputs should work fine as long as we are sure that the input is scalar. I will test and push if it doesn't break the execution of any of the targeted models.

File: src/frontends/pytorch/src/op/arange.cpp (hunk 6 above)
Comment (cavusmustafa) [reply src/frontends/pytorch/src/op/arange.cpp:115]: We can do the same for start and step as well. But it seems like the unit test for arange already have an issue that torch dynamo removes this op. It never reaches to fx_openvino as a single op. Need to fix this issue before updating the test.

**Comment by suryasidd** [comment]
build_jenkins

File: src/frontends/pytorch/src/op/arange.cpp (hunk 6 above)
Comment (cavusmustafa) [reply src/frontends/pytorch/src/op/arange.cpp]: Arange op with dynamic inputs does not seem to be fully  supported in torch dynamo. So, it is excluded from the graph and never gets into OV in the current unit tests. So, this seems to be an issue with the existing arange unit test now (could be observed in OV master as well). I didn't observe this issue for arange ops as part of Llama2. It seems like we need an actual fix for the current issue before adding a new test for the updates from this PR. Do we need this to be resolved in this PR?

**Comment by suryasidd** [comment]
build_jenkins

File: src/bindings/python/src/openvino/frontend/pytorch/fx_decoder.py [hunk 11]
```diff
@@ -317,6 +317,13 @@ def _raw_input(self, index):
     def num_of_outputs(self):
         return len(self.outputs())
 
+    def output_list_size(self):
+        max_out_id = -1
+        for user in self.pt_module.users:
+            if "<built-in function getitem>" == str(user.target) and max_out_id < user.args[1]:
+                max_out_id = user.args[1]
+        return max_out_id+1
```
Comment (mvafin) [inline src/bindings/python/src/openvino/frontend/pytorch/fx_decoder.py:325]:
```suggestion
        return max_out_id + 1
```

File: src/frontends/pytorch/src/op/reshape.cpp [hunk 12]
```diff
@@ -31,55 +31,8 @@ OutputVector translate_reshape_fx(const NodeContext& context) {
...
-            } else {
-                FRONT_END_GENERAL_CHECK(
-                    false,
-                    "Cannot support reshape with multiple dynamic dimensions for dynamic input ranks");
-            }
-        }
-
-        auto shape_const = ov::op::v0::Constant::create(element::i32, Shape{num_inputs - 1}, shape_vec);
-        auto reshape = std::make_shared<ov::op::v1::Reshape>(context.get_input(0), shape_const, true);
+        auto concat = concat_list_from_inputs(context, 1, num_inputs);
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/reshape.cpp:34]: Please validate that input is really of type `List`

File: src/frontends/pytorch/src/op/slice_scatter.cpp [hunk 13]
```diff
@@ -26,7 +26,7 @@ OutputVector translate_slice_scatter_fx(const NodeContext& context) {
     ov::Output<ov::Node> dim;
     if (!context.input_is_none(2)) {
         dim = context.get_input(2);
-        if (dim.get_partial_shape().rank().is_dynamic() || dim.get_partial_shape().rank().get_length() == 0) {
+        if (!(dim.get_partial_shape().rank().is_dynamic()) && dim.get_partial_shape().rank().get_length() == 0) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/slice_scatter.cpp:29]: This change is dangerous. This is a common code and by default we do Unsqueeze for dynamic type

Comment (mvafin) [inline src/frontends/pytorch/src/op/slice_scatter.cpp:29]:
```suggestion
        if (dim.get_partial_shape().rank().is_static() && dim.get_partial_shape().rank().get_length() == 0) {
```

File: src/frontends/pytorch/src/op/slice_scatter.cpp [hunk 14]
```diff
@@ -35,7 +35,7 @@ OutputVector translate_slice_scatter_fx(const NodeContext& context) {
     ov::Output<ov::Node> start;
     if (!context.input_is_none(3)) {
         start = context.get_input(3);
-        if (start.get_partial_shape().rank().is_dynamic() || start.get_partial_shape().rank().get_length() == 0) {
+        if (!(start.get_partial_shape().rank().is_dynamic()) && start.get_partial_shape().rank().get_length() == 0) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/slice_scatter.cpp:38]:
```suggestion
        if (start.get_partial_shape().rank().is_static() && start.get_partial_shape().rank().get_length() == 0) {
```

File: src/frontends/pytorch/src/op/slice_scatter.cpp (hunk 1 above)
Comment (mvafin) [inline src/frontends/pytorch/src/op/slice_scatter.cpp:47]:
```suggestion
        if (end.get_partial_shape().rank().is_static() && end.get_partial_shape().rank().get_length() == 0) {
```

File: src/frontends/pytorch/src/op/slice_scatter.cpp [hunk 15]
```diff
@@ -53,7 +53,7 @@ OutputVector translate_slice_scatter_fx(const NodeContext& context) {
     ov::Output<ov::Node> step;
     if (!context.input_is_none(5)) {
         step = context.get_input(5);
-        if (step.get_partial_shape().rank().is_dynamic() || step.get_partial_shape().rank().get_length() == 0) {
+        if (!(step.get_partial_shape().rank().is_dynamic()) && step.get_partial_shape().rank().get_length() == 0) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/slice_scatter.cpp:56]:
```suggestion
        if (step.get_partial_shape().rank().is_static() && step.get_partial_shape().rank().get_length() == 0) {
```

**Review by mvafin — APPROVED** [review]
This PR lacks tests. 
@cavusmustafa Could you create GHA tests for specific models you are working on?
In a separate PR.
